### PR TITLE
UISI test

### PR DESCRIPTION
**Pull request by JorikSchellekens**
Tests for https://github.com/matrix-org/synapse/issues/5441

### CODE REVIEW

**Comment by erikjohnston** [comment]
The tests are broadly good, but the problem with killing one of the servers is that they are used for the entire test run, rather than being set up for each test. I think a good alternative here is to use the sytest server implementation and then we can simulate the server "going down" by having it return e.g. a 503.

**Comment by JorikSchellekens** [comment]
The server is restarted at the end. Does the sytest server implementation emulate users actions such as joining rooms and uploading keys? Otherwise this test requires a lot of mocking.

**Comment by erikjohnston** [comment]
> Does the sytest server implementation emulate users actions such as joining rooms and uploading keys? Otherwise this test requires a lot of mocking.

Yup, look in `test/50federation/` which all use the perl sytest federation thing.

**Comment by JorikSchellekens** [comment]
I've changed it over to the sytest federation server

File: tests/50federation/40devicelists.pl [hunk 1]
```diff
@@ -284,3 +284,132 @@
...
+            room_version => "5",
+         )->then( sub {
+            my ( $room_id ) = @_;
+            $outbound_client->join_room(
+               server_name => $local_server_info->server_name,
+               room_id     => $room_id,
+               user_id     => $outbound_client_user,
+            )
+         })->then( sub {
+            do_request_json_for( $local_user,
```
Comment (erikjohnston) [inline tests/50federation/40devicelists.pl]: Can you just add some comments at the steps here please, e.g. here we do "First we successfully request the keys while the remote server is up", etc.

File: tests/50federation/40devicelists.pl [hunk 2]
```diff
@@ -284,3 +284,132 @@
...
+            }),
+            $inbound_server->await_request_user_keys_query()->then( sub {
+               my ( $req ) = @_;
+               $OUTPUT->diag("responding");
+               $req->respond_json({}, code => 503);
+               Future->done(1)
+            })
+         );
+         Future->wait_any(
+            @respond_503,
```
Comment (erikjohnston) [inline tests/50federation/40devicelists.pl]: Won't `Future->wait_any` return here when one of the above `await_*` returns? I think you probably want to not wait on them and just return the `do_request_json`?

File: tests/05homeserver.pl [hunk 3]
```diff
@@ -4,7 +4,7 @@
 
 my $N_HOMESERVERS = 2;
 
-my @servers;
+our @SERVERS;
```
[the remark this answers is not in the document]
Comment (JorikSchellekens) [reply tests/05homeserver.pl]: Right

File: tests/50federation/40devicelists.pl (hunk 2 above)
Comment (JorikSchellekens) [reply tests/50federation/40devicelists.pl]: Rookie question: if I do so how do I start the two futures I've created using await*?

File: tests/05homeserver.pl [hunk 4]
```diff
@@ -135,4 +135,4 @@
          })
       },
    );
-} 0 .. $N_HOMESERVERS-1;
+} 0 .. $N_HOMEservers-1;
```
Comment (erikjohnston) [inline tests/05homeserver.pl]: Can we just not change this file? :)

File: tests/50federation/40devicelists.pl [hunk 5]
```diff
@@ -284,3 +284,125 @@
...
+         );
+         do_request_json_for( $local_user,
+            method  => "POST",
+            uri     => "/r0/keys/query",
+            content => {
+               device_keys => {
+                  $outbound_client_user => {}
+               }
+            }
+         )->then( sub {
```
Comment (erikjohnston) [inline tests/50federation/40devicelists.pl]: I would probably move this up a level so that we don't have nested callbacks

**Pull request merged**